### PR TITLE
feat: partial implements gas report for ethereum watcher service

### PR DESCRIPTION
**Pull request by nnmax**

### CODE REVIEW

**Review by rickliujh — COMMENTED** [review]
1. files' name should meaningful, e.g. align with domain
2. how do you manage the state change after? do you need a state management class or state machine?
3. timeout is mandatory if pushNotification is specified. should we need any sanity check in controller in order to prevent any unexpected behaviour from service?

**Comment by rickliujh** [comment]
I have made some update in docs in terms of target's design. you might need to know before merge this pr.
https://github.com/SlowlyCloud/docs/pull/2

File: src/ethereumWatcher/ethereumWatcher.service.ts [hunk 1]
```diff
@@ -1,12 +1,121 @@
...
+            reportData.gasPrice = highestGasTransaction.gasPrice;
+            // TODO: get gasLimit from pending transaction
+            reportData.gasLimit = '';
+            reportData.maxPriorityFeePerGas =
+                highestGasTransaction.maxPriorityFeePerGas;
+            reportData.maxFeePerGas = highestGasTransaction.maxFeePerGas;
+            // TODO: get baseFee from pending transaction
+            reportData.baseFee = '';
+            gasReport.gasReport = reportData;
+            gasReport.pushNotification = data.args.pushNotification;
```
Comment (rickliujh) [inline src/ethereumWatcher/ethereumWatcher.service.ts]: you might want to apply builder pattern here to assign the DTO

File: src/ethereumWatcher/rest/ethereumWatcher.controller.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    @Post('gas-report')
+    @OtelMethodCounter()
+    async report(@Body() data: GasReportRequestBody) {
+        const highestGasReport =
+            await this.ethereumWatcherService.getHighestGasReport(data);
+
+        return Response.new(200, 'successful', highestGasReport);
+    }
+
+    @Delete('gas-report/:id')
```
Comment (rickliujh) [inline src/ethereumWatcher/rest/ethereumWatcher.controller.ts]: @Post()
@Delete(':id')

are able to replace to for simplify

and @Controller can be change below accordingly

@Controller("v1/gas-report")

File: src/ethereumWatcher/queries/transactions.handle.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { Logger } from '@nestjs/common';
+import { QueryHandler } from '@nestjs/cqrs';
+import { Alchemy, Network } from 'alchemy-sdk';
+import { TransactionsQuery } from './transactions.query';
+import type { IQueryHandler } from '@nestjs/cqrs';
+
+@QueryHandler(TransactionsQuery)
+export class TransactionsHandler implements IQueryHandler<TransactionsQuery> {
+    private readonly alchemy = new Alchemy({
```
Comment (rickliujh) [inline src/ethereumWatcher/queries/transactions.handle.ts]: normally, the assign of variable with construction should happen in the constructer of class itself or through injection from dependency container.

File: src/ethereumWatcher/queries/transactions.handle.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    constructor(private readonly logger: Logger) {}
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    async execute(query: TransactionsQuery) {
+        this.logger.verbose(
+            'TransactionsHandler::execute:' + JSON.stringify(query),
+        );
+
+        this.logger.verbose(
+            'InquiryApeHandle::execute:document:' + JSON.stringify(document),
```
Comment (rickliujh) [inline src/ethereumWatcher/queries/transactions.handle.ts]: code that commit at least should able to pass the compiler

File: src/ethereumWatcher/rest/ethereumWatcher.dto.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { Type } from 'class-transformer';
+import {
+    IsBoolean,
+    IsEthereumAddress,
+    IsPositive,
+    IsString,
+} from 'class-validator';
+import { IsInt } from 'class-validator';
+
+class Target {
```
Comment (rickliujh) [inline src/ethereumWatcher/rest/ethereumWatcher.dto.ts]: the Target class doesn't follow the struct defined in https://github.com/SlowlyCloud/docs/pull/2

File: src/ethereumWatcher/rest/ethereumWatcher.dto.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    @IsString()
+    tokenId: string; // tokenId for the specific ethereum address (aka collection/contract address)
+}
+
+export class Args {
+    @IsBoolean()
+    pushNotification: boolean; // specify whether push notification is needed
+
+    @Type(() => Number)
```
Comment (rickliujh) [inline src/ethereumWatcher/rest/ethereumWatcher.dto.ts]: is able to remove this line? it should work just fine with two assertion below.

File: src/ethereumWatcher/ethereumWatcher.service.ts [hunk 7]
```diff
@@ -1,12 +1,121 @@
...
+    }
+
+    private collectPendingTransactions(data: GasReportRequestBody) {
+        const pendingTransactions: PendingTransaction[] = [];
+        const event = {
+            method: AlchemySubscription.PENDING_TRANSACTIONS,
+            toAddress: SEA_PORT_ADDRESS,
+        };
+
+        const pendingTransactionsListener = (
```
Comment (rickliujh) [inline src/ethereumWatcher/ethereumWatcher.service.ts]: is possible keep this function as the method of class in order to reduce this memory allocate from every time calling this function due to the repeated creation of same listener?

**Review by rickliujh — CHANGES_REQUESTED** [review]

**Comment by rickliujh** [comment]
> * files' name should meaningful, e.g. align with domain
> * how do you manage the state change after? do you need a state management class or state machine?
> * timeout is mandatory if pushNotification is specified. should we need any sanity check in controller in order to prevent any unexpected behaviour from service?

and please don't forget to solve these as well. 

Thank you for PR

**Comment by nnmax** [comment]
> 1. files' name should meaningful, e.g. align with domain
> 
>     2. how do you manage the state change after? do you need a state management class or state machine?
> 
>     3. timeout is mandatory if pushNotification is specified. should we need any sanity check in controller in order to prevent any unexpected behaviour from service?

Sorry, I don't quite understand the first and second points.

All other issues have been fixed.

File: src/ethereumWatcher/websocket/ethereumWatcher.websocket.ts [hunk 8]
```diff
@@ -8,8 +8,8 @@ export class EthereumWatcherWebsocketGateway {
         private readonly logger: Logger,
     ) {}
 
-    todo(): void {
+    pendingTransaction(): void {
         this.logger.debug(`EthereumWatcherWebsocketGateway::todo`);
-        this.ws.server.emit('EW:TODO', {});
+        this.ws.server.emit('EW:HIGHEST_GAS', {});
```
Comment (rickliujh) [inline src/ethereumWatcher/websocket/ethereumWatcher.websocket.ts]: I think the name of event GAS_REPORT is better for this case

File: src/ethereumWatcher/rest/ethereumWatcher.controller.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import { OtelMethodCounter } from 'nestjs-otel';
+import { CommandBus, QueryBus } from '@nestjs/cqrs';
+import { InjectMapper } from '@automapper/nestjs';
+import { Mapper } from '@automapper/core';
+import { Response } from 'src/share/rest/dto/response.dto';
+import { EthereumWatcherService } from '../ethereumWatcher.service';
+import { GasReportRequestBody } from './ethereumWatcher.dto';
+
+@Controller('v1/gas-report')
+export class EthereumWatcherController {
```
Comment (rickliujh) [inline src/ethereumWatcher/rest/ethereumWatcher.controller.ts]: What do you think to name this controller to "GasReportController"?

**Comment by rickliujh** [comment]
> > 1. files' name should meaningful, e.g. align with domain
> >    
> >    2. how do you manage the state change after? do you need a state management class or state machine?
> >    3. timeout is mandatory if pushNotification is specified. should we need any sanity check in controller in order to prevent any unexpected behaviour from service?
> 
> Sorry, I don't quite understand the first and second points.
> 
> All other issues have been fixed.

Sounds good. 

By what I mean in first one "meaningful" is to name a file with respects to its domain logic, so it will clear for other developers to read and understand. The file `ethereumWatcher.controller.ts` might feel more concise as the name `gasReport.controller.ts` than just repeating the service name, for example.

For the second one, I'm suggesting that we should maintain the state of transaction (by doing so, we should store the state in DB), so we will know how the gas change from time to time, like the gas growth from 12eth 2023-12-31 to 14eth 2024-1-1 in the sequence

**Review by rickliujh — COMMENTED** [review]
shall we change name of ethereumWatcher.service.ts as well

**Comment by nnmax** [comment]
> shall we change name of ethereumWatcher.service.ts as well

done

**Review by rickliujh — APPROVED** [review]

**Pull request merged**